x/staking/keeper/validator.go

```diff
@@ -358,112 +358,118 @@ func (k Keeper) GetLastValidators(ctx sdk.Context) (validators []types.Validator
-}
-
-// Returns a concatenated list of all the timeslices before currTime, and deletes the timeslices from the queue
-func (k Keeper) GetAllMatureValidatorQueue(ctx sdk.Context, currTime time.Time) (matureValsAddrs []sdk.ValAddress) {
-	// gets an iterator for all timeslices from time 0 until the current Blockheader time
-	validatorTimesliceIterator := k.ValidatorQueueIterator(ctx, ctx.BlockHeader().Time)
-	defer validatorTimesliceIterator.Close()
-
-	for ; validatorTimesliceIterator.Valid(); validatorTimesliceIterator.Next() {
-		timeslice := sdk.ValAddresses{}
-		k.cdc.MustUnmarshalBinaryBare(validatorTimesliceIterator.Value(), &timeslice)
-
-		matureValsAddrs = append(matureValsAddrs, timeslice.Addresses...)
-	}
-
-	return matureValsAddrs
+	return store.Iterator(types.ValidatorQueueKey, sdk.InclusiveEndBytes(types.GetValidatorQueueKey(endTime, endHeight)))
 }
 
-// Unbonds all the unbonding validators that have finished their unbonding period
-func (k Keeper) UnbondAllMatureValidatorQueue(ctx sdk.Context) {
+// UnbondAllMatureValidators unbonds all the mature unbonding validators that
+// have finished their unbonding period.
+func (k Keeper) UnbondAllMatureValidators(ctx sdk.Context) {
 	store := ctx.KVStore(k.storeKey)
 
-	validatorTimesliceIterator := k.ValidatorQueueIterator(ctx, ctx.BlockHeader().Time)
-	defer validatorTimesliceIterator.Close()
-
-	for ; validatorTimesliceIterator.Valid(); validatorTimesliceIterator.Next() {
-		timeslice := sdk.ValAddresses{}
-		k.cdc.MustUnmarshalBinaryBare(validatorTimesliceIterator.Value(), &timeslice)
-
-		for _, valAddr := range timeslice.Addresses {
-			val, found := k.GetValidator(ctx, valAddr)
-			if !found {
-				panic("validator in the unbonding queue was not found")
-			}
+	blockTime := ctx.BlockTime()
+	blockHeight := ctx.BlockHeight()
+
+	// unbondingValIterator will contains all validator addresses indexed under
+	// the ValidatorQueueKey prefix. Note, the entire index key is composed as
+	// ValidatorQueueKey | timeBzLen (8-byte big endian) | timeBz | heightBz (8-byte big endian),
+	// so it may be possible that certain validator addresses that are iterated
+	// over are not ready to unbond, so an explicit check is required.
+	unbondingValIterator := k.ValidatorQueueIterator(ctx, blockTime, blockHeight)
+	defer unbondingValIterator.Close()
+
+	for ; unbondingValIterator.Valid(); unbondingValIterator.Next() {
+		key := unbondingValIterator.Key()
+		keyTime, keyHeight, err := types.ParseValidatorQueueKey(key)
+		if err != nil {
+			panic(fmt.Errorf("failed to parse unbonding key: %w", err))
+		}
 
-			if !val.IsUnbonding() {
-				panic("unexpected validator in unbonding queue; status was not unbonding")
+		// All addresses for the given key have the same unbonding height and time.
+		// We only unbond if the height and time are less than the current height
+		// and time.
+		if keyHeight <= blockHeight && (keyTime.Before(blockTime) || keyTime.Equal(blockTime)) {
```

Comment:
If we iterate based on time (ordered lexicographic ally first), then the time checks should be unnecessary (though they won't break anything)

Reply:
Indeed that is correct, but it costs nothing and provides a clearer context to future readers of this code. Do you agree?

Reply:
Sure, it's fine.